hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpression;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Implements evaluation of SQL comparison predicates.
+ *
+ * @see ComparisonMode
+ */
+public final class ComparisonPredicate extends BiExpression<Boolean> implements IdentifiedDataSerializable {
+
+    private ComparisonMode mode;
+
+    public ComparisonPredicate() {
+        // No-op.
+    }
+
+    private ComparisonPredicate(Expression<?> left, Expression<?> right, ComparisonMode mode) {
+        super(left, right);
+        this.mode = mode;
+    }
+
+    public static ComparisonPredicate create(Expression<?> left, Expression<?> right, ComparisonMode comparisonMode) {
+        assert left.getType().equals(right.getType());
```

Comment:
This assertion fails for some types, e.g. for `QueryDataType DECIMAL_BIG_INTEGER`

When we do `bigIntegerColumn=?`, then on the left side we have `QueryDataType DECIMAL_BIG_INTEGER`, but on the right side we have `QueryDataType DECIMAL`. They are not equal.

Perhaps the correct assert should check for type family, not type.

Reply:
There are no non-canonical types on expressions level: types of literals, parameters and subexpressions are assigned by Calcite side and they are always canonical since Calcite side doesn't know anything about non-canonical ones, `ColumnExpression` is always reporting canonical types now.

Reply:
Consider the following query that fails due to this assertion:
```
SELECT * FROM table WHERE big_integer_column=1
```
This is a valid query with a supported data type. If I change the assertion to 
```
assert left.getType().getTypeFamily().equals(right.getType().getTypeFamily());
```
... then the query works fine and produces correct results. 

The same problem is applicable to `java.lang.Character` columns that are mapped to `VARCHAR_CHARACTER`. In future it will also be a problem for temporal types.

Should we fix it to the assertion I mentioned above?

Reply:
Or maybe this works differently in this PR comparing to the `sql` branch? I tried it in the `sql` branch. So perhaps I reported the problem that doesn't exist anymore.

Reply:
yes, just checked it, it works fine against this PR branch